File: src/main/java/me/rsman/BetterMinecraftCore/Listeners/EnchantListener.java
package me.rsman.BetterMinecraftCore.Listeners;

import me.rsman.BetterMinecraftCore.BetterMinecraftCore;
import me.rsman.BetterMinecraftCore.Enchantments.CustomEnchantClass;
import me.rsman.BetterMinecraftCore.Managers.EnchantManager;
import me.rsman.BetterMinecraftCore.Managers.ItemManager;
import me.rsman.BetterMinecraftCore.Managers.ItemTypeChecker;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.enchantments.EnchantmentOffer;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.enchantment.EnchantItemEvent;
import org.bukkit.event.enchantment.PrepareItemEnchantEvent;

import java.util.*;

public class EnchantListener implements Listener {
    @EventHandler
    public void addCustomEnchantFromTable(EnchantItemEvent e){
        Map<Enchantment,Map<Enchantment, Integer>> enchantsToReplace = new HashMap<>();
        for (Map.Entry<Enchantment, Integer> enchantToAdd: e.getEnchantsToAdd().entrySet()) {
           if(enchantToAdd.getKey().getKey().toString().equals("minecraft:protection")){
               enchantsToReplace.put(enchantToAdd.getKey(), new HashMap<Enchantment, Integer>(){{put(EnchantManager.enchants.get("minecraft:bmc_protection"),enchantToAdd.getValue());}});
           }
        }
        for (Map.Entry<Enchantment,Map<Enchantment, Integer>> enchantToReplace: enchantsToReplace.entrySet()) {
            e.getEnchantsToAdd().remove(enchantToReplace.getKey());
            Map.Entry<Enchantment, Integer> enchant = enchantToReplace.getValue().entrySet().iterator().next();
            e.getEnchantsToAdd().put(enchant.getKey(), enchant.getValue());
        }
        List<CustomEnchantClass> applicableEnchants = new ArrayList<>();
        for(Map.Entry<String, Enchantment> customEnchant : EnchantManager.enchants.entrySet()){
            if(customEnchant.getValue() instanceof CustomEnchantClass){
                if(((CustomEnchantClass) customEnchant.getValue()).isApplicable(e.getItem()) && ((CustomEnchantClass) customEnchant.getValue()).getMinimumLevel() < e.getExpLevelCost()){
                    applicableEnchants.add((CustomEnchantClass)customEnchant.getValue());
                }
            }
        }
        while(Math.random() > 0.5 && applicableEnchants.size() > 0){
            int enchantKey = (int)Math.round(Math.random() * (applicableEnchants.size()-1));
            int deltaLv = e.getExpLevelCost() - applicableEnchants.get(enchantKey).getMinimumLevel();
            int deltaMax = 30 - applicableEnchants.get(enchantKey).getMinimumLevel();
            int level = Math.max(1,
                            Math.min(applicableEnchants.get(enchantKey).getMaxLevel(),
                                Math.round((float)(deltaLv/deltaMax)*applicableEnchants.get(enchantKey).getMaxLevel())
                            )
                        );
            e.getEnchantsToAdd().put(applicableEnchants.get(enchantKey), level);
            applicableEnchants.remove(enchantKey);
        }
        for (Map.Entry<Enchantment, Integer> enchantToAdd: e.getEnchantsToAdd().entrySet()) {
            e.getItem().addUnsafeEnchantment(enchantToAdd.getKey(), enchantToAdd.getValue());
        }
        BetterMinecraftCore.getInstance().getLogger().info(e.getEnchantsToAdd().toString());
        ItemManager.updateItemLore(e.getItem());
    }
}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/Listeners/EnchantListener.kt
package me.rsman.BetterMinecraftCore.Listeners

import java.util.HashMap
import org.bukkit.enchantments.Enchantment
import me.rsman.BetterMinecraftCore.enums.EEnchants
import me.rsman.BetterMinecraftCore.Managers.ItemManager
import me.rsman.BetterMinecraftCore.Enchantments.CustomEnchantClass
import me.rsman.BetterMinecraftCore.extensions.updateCustomLore
import org.bukkit.event.enchantment.EnchantItemEvent
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import java.util.ArrayList

class EnchantListener : Listener {
    @EventHandler
    fun addCustomEnchantFromTable(e: EnchantItemEvent) {
        val enchantsToReplace = mutableMapOf<Enchantment, Map<Enchantment, Int>>()
        for ((key, value) in e.enchantsToAdd) {
            if (EEnchants.replacesMapFromNamespaces.containsKey(key.key.toString())) {
                enchantsToReplace[key] = mapOf(EEnchants.replacesMapFromNamespaces[key.key.toString()]!! to value)
            }
        }
        for ((key, value) in enchantsToReplace) {
            e.enchantsToAdd.remove(key)
            val (key1, value1) = value.entries.iterator().next()
            e.enchantsToAdd[key1] = value1
        }
        val applicableEnchants: MutableList<CustomEnchantClass> = ArrayList()
        for (ench in EEnchants.values()) {
            if (ench.enchant is CustomEnchantClass) {
                if (ench.enchant.isApplicable(e.item) && ench.enchant.minimumLevel < e.expLevelCost) {
                    applicableEnchants.add(ench.enchant)
                }
            }
        }
        while (Math.random() > 0.5 && applicableEnchants.size > 0) {
            val enchantKey = Math.round(Math.random() * (applicableEnchants.size - 1)).toInt()
            val deltaLv = e.expLevelCost - applicableEnchants[enchantKey].minimumLevel
            val deltaMax = 30 - applicableEnchants[enchantKey].minimumLevel
            val level = Math.max(1,
                    Math.min(applicableEnchants[enchantKey].maxLevel,
                            Math.round(deltaLv.toFloat() / deltaMax.toFloat() * applicableEnchants[enchantKey].maxLevel)
                    )
            )
            e.enchantsToAdd[applicableEnchants[enchantKey]] = level
            applicableEnchants.removeAt(enchantKey)
        }
        for ((key, value) in e.enchantsToAdd) {
            e.item.addUnsafeEnchantment(key!!, value!!)
        }
        e.item.updateCustomLore()
    }
}